src/libostree/ostree-repo.c

```diff
@@ -405,26 +412,40 @@ pop_repo_lock (OstreeRepo  *self,
   GHashTable *lock_table = g_private_get (&repo_lock_table);
   g_return_val_if_fail (lock_table != NULL, FALSE);
 
-  OstreeRepoLock *lock = g_hash_table_lookup (lock_table, self);
-  g_return_val_if_fail (lock != NULL, FALSE);
-  g_return_val_if_fail (lock->fd != -1, FALSE);
+  GPtrArray *stack = g_hash_table_lookup (lock_table, self);
+  g_return_val_if_fail (stack != NULL, FALSE);
+
+  g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&self->lock.mutex);
+  g_return_val_if_fail (self->lock.fd != -1, FALSE);
 
   OstreeRepoLockInfo info;
-  repo_lock_info (lock, &info);
+  repo_lock_info (self, &info);
   g_return_val_if_fail (info.len > 0, FALSE);
 
   g_debug ("Pop lock: state=%s, depth=%u", info.name, info.len);
+  int state_to_drop = GPOINTER_TO_INT (stack->pdata[stack->len - 1]);
```

Comment:
I think here there is a missing check to prevent `stack->len - 1` from underflowing, which I think can happen in case of unbalanced `unlock()` calls from a consuming thread that previously performed at least one `lock()` call.

Reply:
Good call, I think that is something that could happen.

Reply:
I stopped using the stack, but the same principle applied to the counters, so I added a check for that.